File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/core/CoreNotification.kt
package dev.bota.sdk.internal.core

import dev.bota.sdk.internal.jni.NativeCoreException
import dev.bota.sdk.internal.jni.NativePacket

internal enum class CoreNotificationKind(val wireValue: Int) {
    Started(0x0401),
    DeviceDiscovered(0x0402),
    ConnectionEstablished(0x0403),
    Progress(0x0404),
    Retrying(0x0405),
    DeviceUploadPreserved(0x0406),
    BleFallbackReady(0x0407),
    FirmwareProgress(0x0408),
    DeviceLog(0x0409),
    StreamingPaused(0x040d),
    StreamingResumed(0x040e),
    StreamingCompleted(0x040f),
    Completed(0x040a),
    Cancelled(0x040b),
    Failed(0x040c),
    ;

    val isTerminal: Boolean get() = this == Completed || this == Cancelled || this == Failed

    companion object {
        fun fromWireValue(value: Int): CoreNotificationKind = entries.firstOrNull { it.wireValue == value }
            ?: throw NativeCoreException(3, 0, false, -1, "unknown notification kind $value")
    }
}

internal data class CoreNotification(val kind: CoreNotificationKind, val packet: NativePacket) {
    val isTerminal: Boolean get() = kind.isTerminal

    companion object {
        fun fromPacket(packet: NativePacket): CoreNotification =
            CoreNotification(CoreNotificationKind.fromWireValue(packet.kind), packet)
    }
}

File: platforms/apple/Sources/BotaAppleSDK/Core/CoreNotification.swift
import BotaDeviceSDKC

enum CoreNotificationKind: Equatable, Sendable {
    case started
    case deviceDiscovered
    case connectionEstablished
    case progress
    case retrying
    case deviceUploadPreserved
    case bleFallbackReady
    case firmwareProgress
    case deviceLog
    case streamingPaused
    case streamingResumed
    case streamingCompleted
    case completed
    case cancelled
    case failed
}

struct CoreNotification: Equatable, Sendable {
    let kind: CoreNotificationKind
    let packet: CorePacket

    var isTerminal: Bool {
        kind == .completed || kind == .cancelled || kind == .failed
    }

    init(packet: CorePacket) throws {
        let kind: CoreNotificationKind
        switch packet.kind {
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_STARTED): kind = .started
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_DEVICE_DISCOVERED): kind = .deviceDiscovered
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_CONNECTION_ESTABLISHED): kind = .connectionEstablished
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_PROGRESS): kind = .progress
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_RETRYING): kind = .retrying
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_DEVICE_UPLOAD_PRESERVED): kind = .deviceUploadPreserved
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_BLE_FALLBACK_READY): kind = .bleFallbackReady
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_FIRMWARE_PROGRESS): kind = .firmwareProgress
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_DEVICE_LOG): kind = .deviceLog
        case 0x040d: kind = .streamingPaused
        case 0x040e: kind = .streamingResumed
        case 0x040f: kind = .streamingCompleted
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_COMPLETED): kind = .completed
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_CANCELLED): kind = .cancelled
        case UInt32(BOTA_DEVICE_SDK_V1_NOTIFICATION_FAILED): kind = .failed
        default:
            throw CoreError(
                code: UInt32(BOTA_DEVICE_SDK_V1_ERROR_UNKNOWN_PACKET),
                operation: packet.operation,
                retryable: false,
                protocolStatus: nil,
                detail: "unknown notification kind \(packet.kind)"
            )
        }
        self.kind = kind
        self.packet = packet
    }
}
